File: Numeros Pseudo-Aleatorios/Multiply With Carry/MultiplyWithCarry.dart
/*
 PRNG .......: Multiply With Carry (MWC)
 Ano ........: Desconhecido
 Autor ......: George Marsaglia
*/

import 'dart:io';

class CmwcRnd {
	
  List<int> _randomValues = List<int>.filled(4096, 0);
  int _carry = 0;
  int _currentIndex = 0;

  int _log2ibase(int n, int acc) {
    if (n == 1)
      return acc;
    else
      return _log2ibase(n >> 1, acc + 1);
  }

  int _log2i(int n) {
    return _log2ibase(n, 0);
  }

  int _rand32() {
    int t;
    int x;
    _currentIndex = (_currentIndex + 1) & 4095;
    t = 18782 * _randomValues[_currentIndex] + _carry;
    _carry = (t >> 32);
    x = t + _carry;
    if (x < _carry) {
      x++;
      _carry++;
    }
    return _randomValues[_currentIndex] = 0xFFFFFFFF - x;
  }

  void initCMWC(int seed) {
    int i;
    for (i = 0; i <= 4095; i++) {
      seed = (seed * 1103515245 + 12345) & 0xFFFFFFFF;
      _randomValues[i] = seed;
    }
    do {
      seed = (seed * 1103515245 + 12345) & 0xFFFFFFFF;
      _carry = seed;
    } while (!(_carry < 809430660));
    _currentIndex = 4095;
  }

  int randCMWC() {
    return _rand32() & 0x7FFFFFFF; // Máscara de bits para descartar o bit de sinal
  }
  
}

void main() {
	
  CmwcRnd cmwc1 = CmwcRnd();
  CmwcRnd cmwc2 = CmwcRnd();
  
  int Xn1 = 1234;
  int Xn2 = DateTime.now().millisecondsSinceEpoch;

  cmwc1.initCMWC(Xn1);
  cmwc2.initCMWC(Xn2);

  print('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  print('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  print('PADRAO: Multiply With Carry (MWC)');
  print('==========================================\n');
  print('    Semente: 1234        Semente: TIME');
  print('    -------------        -------------');

  for (int i = 1; i <= 10; i++) {
    print('${cmwc1.randCMWC().toString().padLeft(17)}${cmwc2.randCMWC().toString().padLeft(21)}');
  }
  
}

File: Numeros Pseudo-Aleatorios/Multiply With Carry/MultiplyWithCarry.cs
/*
 PRNG .......: Multiply With Carry (MWC)
 Ano ........: Desconhecido
 Autor ......: George Marsaglia
*/

using System;

public class CmwcRnd
{
	
    private int[] randomValues;
    private int carry;
    private int currentIndex;

    private int log2ibase(int n, int acc)
    {
        if (n == 1)
            return acc;
        else
            return log2ibase(n >> 1, acc + 1);
    }

    private int log2i(int n)
    {
        return log2ibase(n, 0);
    }

    public CmwcRnd()
    {
        randomValues = new int[4096];
        currentIndex = 0;
    }

    public void initCMWC(int seed)
    {
        for (int i = 0; i <= 4095; i++)
        {
            seed = (int)((seed * 1103515245L + 12345) & 0xFFFFFFFF);
            randomValues[i] = seed;
        }

        do
        {
            seed = (int)((seed * 1103515245L + 12345) & 0xFFFFFFFF);
            carry = seed;
        } while (!(carry < 809430660));

        currentIndex = 4095;
    }

    public int randCMWC()
    {
        long t;
        long x;
        currentIndex = (currentIndex + 1) & 4095;
        t = 18782L * randomValues[currentIndex] + carry;
        carry = (int)(t >> 32);
        x = t + carry;
        if (x < carry)
        {
            x++;
            carry++;
        }
        return (int)(x & 0x7FFFFFFF); // Retorna valor positivo
    }

    public static void Main(string[] args)
    {
		
        CmwcRnd cmwc1 = new CmwcRnd();
        CmwcRnd cmwc2 = new CmwcRnd();
        
        int Xn1 = 1234;
        int Xn2 = (int)DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        cmwc1.initCMWC(Xn1);
        cmwc2.initCMWC(Xn2);

        Console.WriteLine("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)");
        Console.WriteLine("PSEUDORANDOM NUMBER GENERATOR (PRNG)");
        Console.WriteLine("PADRAO: Multiply With Carry (MWC)");

        Console.WriteLine("\n    Semente: 1234        Semente: TIME");
        Console.WriteLine("    -------------        -------------");

        for (int i = 1; i <= 10; i++)
        {
            Console.WriteLine($"{cmwc1.randCMWC(),17}           {cmwc2.randCMWC(),10}");
        }
        
    }
    
}
